### PR TITLE
Fix bugs in dependencies, improve tests

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-226

Changes:
- Cleanup in dependencies hierarchy.
- Improved mocks, mocked dependencies are similar to production dependencies as much as possible.
  - This reduces the probability of a bug that we do not catch during testing.
- Aded new common scopes `RequestInfo`, `EtcdClientScope`, `TaskScope`, `DistributionScope`, `OrchestratorScope`.
- Update documentation.
- Improved setup and cleanup of temporary etcd namespace for a test.
- Removed bugs found during work.

-----

Je to dost vela zmien `3k`, vacsina su:
- Opravy namingu.
- Rozdelenie kodu do viacerych suborov.
- Doplnene testy.
- Zmeny v generovanom kode.
- Okomentovane po commitoch.
- Podarilo sa mi z toho vyclenit do samostatneho PR iba https://github.com/keboola/keboola-as-code/pull/1359, kedze ked sa zmenia dependencies musi sa upravit aj kod.
- Pozri prosim najprv novu dokumentaciu nizsie.
- A idealne si asi aj preklikat ako vyzera vysledny kod:
  - Malo by to byt jednoduchsie na citanie/pochopenie, ako stara verzia.
  - [common](https://github.com/keboola/keboola-as-code/tree/michaljurecko-PSGO-226-common-deps/internal/pkg/service/common/dependencies)
  - [templates](https://github.com/keboola/keboola-as-code/tree/michaljurecko-PSGO-226-common-deps/internal/pkg/service/templates/dependencies)
  - [buffer](https://github.com/keboola/keboola-as-code/tree/michaljurecko-PSGO-226-common-deps/internal/pkg/service/buffer/dependencies)
  - [CLI](https://github.com/keboola/keboola-as-code/tree/michaljurecko-PSGO-226-common-deps/internal/pkg/service/cli/dependencies)
  

-----
#### Common Dependencies

image

#### Templates Service

image

#### Buffer Service

image

#### CLI Service

image

### CODE REVIEW

File: test/buffer/api/import/import/expected-server-stderr [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+[bufferApi][http][requestId=%s]INFO failed to format record for export "export-1":
+- path "col1" not found in the body [%s] (errors.withStatusCode):
+- failed to format record for export "export-1":
+- path "col1" not found in the body [%s] (errors.BadRequestError):
+  - failed to format record for export "export-1" [%s]:
+    - path "col1" not found in the body [%s] (*errors.wrappedError):
+      - jsonnet error: RUNTIME ERROR: path "col1" not found in the body [%s] (*fmt.wrapError):
+        - RUNTIME ERROR: path "col1" not found in the body | errorName=buffer.badRequest errorType=errors.withStatusCode response=%s
+[bufferApi]INFO exiting (terminated)
+%A
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply test/buffer/api/import/import/expected-server-stderr:16]: Nemozem tu dat uplne cely stdout, lebo niektore operacie nemaju pevne poradie v logoch.
Napr. logy s prefixom `[bufferApi][api][watcher]` a `[bufferApi][task]` mozu prist v nahodom poradi, po skonceni tasku.

Cely vystup vyzera takto:

<details>
  <summary>Click me</summary>
<pre>
[bufferApi]INFO Configuration: debug-log=false; debug-etcd=false; debug-http=false; datadog-enabled=false; datadog-debug=false; cpu-profile=; storage-api-host=connection.keboola.com; etcd-endpoint=etcd:2379; etcd-namespace=3D1T62qwoCvp6SwfBoLy/; etcd-username=root; etcd-connect-timeout=30s; unique-id=; public-address=https://buffer.keboola.local; listen-address=localhost:46183; metrics-listen-address=localhost:37669; statistics-sync-interval=1s; receiver-buffer-size=50MB; receiver-buffer-size-cache-ttl=1s;
[bufferApi]INFO process unique id "23bb6df250aa-159531"
[bufferApi][metrics]INFO HTTP server listening on "localhost:37669/metrics"
[bufferApi]INFO loading Storage API index without components
[bufferApi]INFO loaded Storage API index without components | 763.951946ms
[bufferApi][etcd-client]INFO connecting to etcd, connectTimeout=30s, keepAliveTimeout=5s, keepAliveInterval=10s
[bufferApi][etcd-client]INFO connected to etcd cluster "etcd:2379" | 115.615082ms
[bufferApi][task][etcd-session]INFO creating etcd session
[bufferApi][task][etcd-session]INFO created etcd session | 86.690699ms
[bufferApi][stats-cache]INFO initialized | 119.136129ms
[bufferApi][api][watcher][etcd-session]INFO creating etcd session
[bufferApi][api][watcher][etcd-session]INFO created etcd session | 61.538648ms
[bufferApi][api][watcher]INFO reported revision "1"
[bufferApi][api][watcher]INFO initialized | 71.118383ms
[bufferApi]INFO starting Buffer API HTTP server, listen-address=localhost:46183
[bufferApi]INFO mounted HTTP endpoints
[bufferApi]INFO HTTP server listening on "localhost:46183"
[bufferApi][task][9483/receiver-1/receiver.create/2023-06-23T12:55:53.295Z_8VEbA]INFO started task
[bufferApi][http][requestId=YytJPWDHKRE5vqR]INFO req /v1/receivers status=202 bytes=276 time=428.370714ms client_ip=127.0.0.1 agent=go-resty/2.7.0 (https://github.com/go-resty/resty)
[bufferApi][http][requestId=e2nP4wH0TeW4pFj]INFO req /v1/tasks/receiver-1/receiver.create/2023-06-23T12:55:53.295Z_8VEbA status=200 bytes=276 time=388.161903ms client_ip=127.0.0.1 agent=go-resty/2.7.0 (https://github.com/go-resty/resty)
[bufferApi][http][requestId=oTGJzI1vN8abQ0K]INFO req /v1/tasks/receiver-1/receiver.create/2023-06-23T12:55:53.295Z_8VEbA status=200 bytes=276 time=335.813007ms client_ip=127.0.0.1 agent=go-resty/2.7.0 (https://github.com/go-resty/resty)
[bufferApi][http][requestId=wPUofSfl6hjmR8p]INFO req /v1/tasks/receiver-1/receiver.create/2023-06-23T12:55:53.295Z_8VEbA status=200 bytes=276 time=361.959061ms client_ip=127.0.0.1 agent=go-resty/2.7.0 (https://github.com/go-resty/resty)
[bufferApi][http][requestId=OBJwJg0R9V4LRfW]INFO req /v1/tasks/receiver-1/receiver.create/2023-06-23T12:55:53.295Z_8VEbA status=200 bytes=276 time=343.396755ms client_ip=127.0.0.1 agent=go-resty/2.7.0 (https://github.com/go-resty/resty)
[bufferApi][api][watcher]INFO created config/receiver/9483/receiver-1
[bufferApi][task][9483/receiver-1/receiver.create/2023-06-23T12:55:53.295Z_8VEbA]INFO task succeeded (11.269704402s): receiver created outputs: {"receiverId":"receiver-1"}
[bufferApi][api][watcher]INFO created slice/active/opened/writing/9483/receiver-1/export-1/2023-06-23T12:55:55.037Z/2023-06-23T12:55:55.037Z
[bufferApi][api][watcher]INFO state updated to the revision "10630"
[bufferApi][api][watcher]INFO reported revision "10630"
[bufferApi][http][requestId=WzldznhhMGZ1A0S]INFO req /v1/tasks/receiver-1/receiver.create/2023-06-23T12:55:53.295Z_8VEbA status=200 bytes=395 time=340.478418ms client_ip=127.0.0.1 agent=go-resty/2.7.0 (https://github.com/go-resty/resty)
[bufferApi][http][requestId=1nzd4uSxe4q9dmA]INFO failed to format record for export "export-1":
- cannot parse request body: unsupported content type "text/plain", supported types: application/json and application/x-www-form-urlencoded [/code/internal/pkg/service/buffer/api/receive/importer.go:142] (errors.withStatusCode):
- failed to format record for export "export-1" [/code/internal/pkg/service/buffer/api/receive/importer.go:117]:
  - cannot parse request body: unsupported content type "text/plain", supported types: application/json and application/x-www-form-urlencoded [/code/internal/pkg/service/buffer/api/receive/receivectx/receivectx.go:64] (*fmt.wrapError):
    - unsupported content type "text/plain", supported types: application/json and application/x-www-form-urlencoded | errorName=buffer.unsupportedMediaType errorType=errors.withStatusCode response={"statusCode":415,"error":"buffer.unsupportedMediaType","message":"Unsupported content type \"text/plain\", supported types: application/json and application/x-www-form-urlencoded."}
[bufferApi][http][requestId=6gSo5EOHgz3Vl7u]INFO failed to format record for export "export-1":
- path "col1" not found in the body [/code/internal/pkg/service/buffer/api/receive/importer.go:142] (errors.withStatusCode):
- failed to format record for export "export-1":
- path "col1" not found in the body [/code/internal/pkg/service/buffer/api/receive/importer.go:124] (errors.BadRequestError):
  - failed to format record for export "export-1" [/code/internal/pkg/service/buffer/api/receive/importer.go:117]:
    - path "col1" not found in the body [/code/internal/pkg/service/buffer/api/receive/jsonnet/jsonnet.go:50] (*errors.wrappedError):
      - jsonnet error: RUNTIME ERROR: path "col1" not found in the body [/code/internal/pkg/encoding/jsonnet/jsonnet.go:46] (*fmt.wrapError):
        - RUNTIME ERROR: path "col1" not found in the body | errorName=buffer.badRequest errorType=errors.withStatusCode response={"statusCode":400,"error":"buffer.badRequest","message":"Failed to format record for export \"export-1\":\n- path \"col1\" not found in the body."}
[bufferApi]INFO exiting (terminated)
[bufferApi]INFO shutting down HTTP server at "localhost:46183"
[bufferApi]INFO HTTP server shutdown finished
[bufferApi][api][watcher]INFO received shutdown request
[bufferApi][api][watcher][etcd-session]INFO closing etcd session
[bufferApi][api][watcher][etcd-session]INFO closed etcd session | 2.517052ms
[bufferApi][api][watcher]INFO shutdown done
[bufferApi][stats]INFO received shutdown request
[bufferApi][stats]INFO shutdown done
[bufferApi][stats-cache]INFO received shutdown request
[bufferApi][stats-cache]INFO shutdown done
[bufferApi][task]INFO received shutdown request
[bufferApi][task][etcd-session]INFO closing etcd session
[bufferApi][task][etcd-session]INFO closed etcd session | 2.223116ms
[bufferApi][task]INFO shutdown done
[bufferApi][etcd-client]INFO closing etcd connection
[bufferApi][etcd-client]INFO closed etcd connection | 222.598µs
[bufferApi][metrics]INFO shutting down HTTP server at "localhost:37669"
[bufferApi][metrics]INFO HTTP server shutdown finished
[bufferApi]INFO exited
</pre>
</details>

File: internal/pkg/service/buffer/worker/service/slice_retry_test.go [hunk 2]
```diff
@@ -104,6 +104,13 @@ func TestRetryFailedUploadsTask(t *testing.T) {
 	_, err := service.New(workerScp)
 	assert.NoError(t, err)
 
+	// Mock events send request
+	workerMock.MockedHTTPTransport().RegisterResponder(
+		http.MethodPost,
+		"https://mocked.transport.http/v2/storage/events",
+		httpmock.NewJsonResponderOrPanic(http.StatusOK, &keboola.Event{ID: "123"}),
+	)
+
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/worker/service/slice_retry_test.go:113]: Ked sa nepodari poslat event je to warning ale lepsie ked tam ziaden warning v logoch nie je. Chybala tato mocked response.

File: internal/pkg/service/templates/api/service/cron.go [hunk 3]
```diff
@@ -17,9 +17,8 @@ func StartRepositoriesPullCron(ctx context.Context, d dependencies.APIScope) err
 	manager := d.RepositoryManager()
 
 	// Start background work
+	d.Logger().Infof("repository pull cron ready")
```
Comment (michaljurecko) [inline internal/pkg/service/templates/api/service/cron.go:20]: Fixnuta stabilita logov, vid test nizsie.

File: test/buffer/api/base/api-startup/expected-server-stderr [hunk 4]
```diff
@@ -0,0 +1,39 @@
+[bufferApi]INFO Configuration: debug-log=false; debug-etcd=false; debug-http=false; datadog-enabled=false; datadog-debug=false; cpu-profile=; storage-api-host=%s; etcd-endpoint=%s; etcd-namespace=%s; etcd-username=%s; etcd-connect-timeout=30s; unique-id=; public-address=https://buffer.keboola.local; listen-address=localhost:%d; metrics-listen-address=localhost:%d; statistics-sync-interval=1s; receiver-buffer-size=50MB; receiver-buffer-size-cache-ttl=1s;
+[bufferApi]INFO process unique id "%s"
+[bufferApi][metrics]INFO HTTP server listening on "localhost:%d/metrics"
+[bufferApi]INFO loading Storage API index without components
```
Comment (michaljurecko) [inline test/buffer/api/base/api-startup/expected-server-stderr:4]: Pridal som testy pre Buffer a Templates API, ktore E2E testuju co sa loguje pri starte API.

Takto sa mozu do buducna odhalit pripadne neplanovane zmeny/chyby v `dependencies`.

Comment (michaljurecko) [reply test/buffer/api/base/api-startup/expected-server-stderr:4]: Na kontroluje je to jednoduche, vsetko sa vypina v opacnom poradi ako sa spustalo:
image

File: internal/pkg/service/buffer/api/dependencies/dependencies.go [hunk 5]
```diff
@@ -1,240 +0,0 @@
...
-	}
-	d.tokenManager = token.NewManager(d)
-	d.tableManager = table.NewManager(d.KeboolaProjectAPI())
-	d.fileManager = file.NewManager(d.Clock(), d.KeboolaProjectAPI(), nil)
-	d.taskNode, err = task.NewNode(d)
-	if err != nil {
-		return nil, err
-	}
-	return d, nil
-}
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/api/dependencies/dependencies.go:184]: :lady_beetle: tu bol bug `d.taskNode, err = task.NewNode(d)`:
- Tj. pre kazdy request authentikovany tokenom sa vytvarala nova `task.Node` + nova etcd session.
  - Nemalo to vplyv na funkcnost ale na performance ano + neporiadok v logoch.

Comment (michaljurecko) [reply internal/pkg/service/buffer/api/dependencies/dependencies.go:184]: Teraz tam uz tato chyba nie je:

https://github.com/keboola/keboola-as-code/blob/315ef855a88c42758de9a142493359712f74b868/internal/pkg/service/buffer/dependencies/dependencies.go#L79-L85

https://github.com/keboola/keboola-as-code/blob/315ef855a88c42758de9a142493359712f74b868/internal/pkg/service/buffer/dependencies/reqproject.go#L36-L53

File: internal/pkg/service/buffer/statistics/cache/cache.go [hunk 6]
```diff
@@ -52,7 +52,7 @@ type Dependencies interface {
 func NewNode(d Dependencies) (*Node, error) {
 	// Create
 	n := &Node{
-		logger: d.Logger().AddPrefix("[stats][cache]"),
+		logger: d.Logger().AddPrefix("[stats-cache]"),
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/statistics/cache/cache.go:55]: Aby sa to dalo v teste/logs odlisit.

File: internal/pkg/service/buffer/watcher/apinode/apinode.go [hunk 7]
```diff
@@ -74,7 +75,7 @@ func New(d Dependencies, opts ...Option) (*Node, error) {
 		clock:  d.Clock(),
 		logger: d.Logger().AddPrefix("[api][watcher]"),
 		client: d.EtcdClient(),
-		stats:  statistics.NewCollectorNode(d),
+		stats:  d.StatsCollector(),
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/watcher/apinode/apinode.go:78]: A tu bol dalsi :lady_beetle: , namiesto nacitania instancie z `dependencies` sa tu vytvarala nova.
- Nemalo to vplyv na funckiu, ale blbo sa debugovalo, preco su vsetky logs 2x.

File: internal/pkg/service/templates/api/config/config.go [hunk 8]
```diff
@@ -20,25 +21,22 @@ const (
...
+	DebugEtcd            bool                   `mapstructure:"debug-etcd" usage:"Enable logging of each etcd KV operation as a debug message."`
+	DebugHTTP            bool                   `mapstructure:"debug-http" usage:"Log HTTP client request and response bodies."`
+	DatadogEnabled       bool                   `mapstructure:"datadog-enabled" usage:"Enable Datadog telemetry integration."`
+	DatadogDebug         bool                   `mapstructure:"datadog-debug" usage:"Enable Datadog debug logs."`
+	CpuProfFilePath      string                 `mapstructure:"cpu-profile" usage:"Write cpu profile to the file."`
+	ListenAddress        string                 `mapstructure:"listen-address" usage:"API HTTP server listen address."`
+	MetricsListenAddress string                 `mapstructure:"metrics-listen-address" usage:"Prometheus /metrics HTTP endpoint listen address."`
+	StorageAPIHost       string                 `mapstructure:"storage-api-host" usage:"Host of the Storage API."`
+	Etcd                 etcdclient.Credentials `mapstructure:",squash" usage:"etcd client credentials."`
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/templates/api/config/config.go:33]: tj. sprava sa to rovnako ako ked tu boli 4 fields, ktore su uz teraz v `Credentials` struct.

**Review by zajca — APPROVED** [review]
Uff, náročné čtení.

**Pull request merged**